### PR TITLE
Merge visitors in AST validation

### PR DESCRIPTION
**Pull request by Zoxc**
Cuts runtime for AST validation on `syntex_syntax` from 31.5 ms to 17 ms.

### CODE REVIEW

File: src/librustc_passes/ast_validation.rs [hunk 1]
```diff
@@ -20,9 +20,79 @@ use errors::Applicability;
 
 struct AstValidator<'a> {
     session: &'a Session,
+
+    // Used to ban nested `impl Trait`, e.g., `impl Into<impl Debug>`.
+    // Nested `impl Trait` _is_ allowed in associated type position,
+    // e.g `impl Iterator<Item=impl Debug>`
```
Comment (Centril) [inline src/librustc_passes/ast_validation.rs]: Maybe also add a note about why we do this (cause we are not sure whether we want it to be higher rank: `impl for<T: Debug> Into<T>`, or rank-1).

File: src/librustc_passes/ast_validation.rs [hunk 2]
```diff
@@ -20,9 +20,79 @@ use errors::Applicability;
...
+    outer_impl_trait: Option<Span>,
+
+    // Used to ban `impl Trait` in path projections like `<impl Iterator>::Item`
+    // or `Foo::Bar<impl Trait>`
+    is_impl_trait_banned: bool,
 }
 
 impl<'a> AstValidator<'a> {
+    fn with_banned_impl_trait<F>(&mut self, f: F)
```
Comment (Centril) [inline src/librustc_passes/ast_validation.rs]: use `f: impl FnOnce(&mut Self)` for increased readability?

File: src/librustc_passes/ast_validation.rs [hunk 3]
```diff
@@ -20,9 +20,79 @@ use errors::Applicability;
...
+        where F: FnOnce(&mut Self)
+    {
+        let old_is_impl_trait_banned = self.is_impl_trait_banned;
+        self.is_impl_trait_banned = true;
+        f(self);
+        self.is_impl_trait_banned = old_is_impl_trait_banned;
+    }
+
+    fn with_impl_trait<F>(&mut self, outer_impl_trait: Option<Span>, f: F)
```
Comment (Centril) [inline src/librustc_passes/ast_validation.rs]: Same.. `f: impl FnOnce(&mut Self)`.

File: src/librustc_passes/ast_validation.rs [hunk 4]
```diff
@@ -20,9 +20,79 @@ use errors::Applicability;
...
+    // Used to ban `impl Trait` in path projections like `<impl Iterator>::Item`
+    // or `Foo::Bar<impl Trait>`
+    is_impl_trait_banned: bool,
 }
 
 impl<'a> AstValidator<'a> {
+    fn with_banned_impl_trait<F>(&mut self, f: F)
+        where F: FnOnce(&mut Self)
+    {
+        let old_is_impl_trait_banned = self.is_impl_trait_banned;
```
Comment (Centril) [inline src/librustc_passes/ast_validation.rs]: Use `mem::replace`?

File: src/librustc_passes/ast_validation.rs [hunk 5]
```diff
@@ -20,9 +20,79 @@ use errors::Applicability;
...
+        self.is_impl_trait_banned = true;
+        f(self);
+        self.is_impl_trait_banned = old_is_impl_trait_banned;
+    }
+
+    fn with_impl_trait<F>(&mut self, outer_impl_trait: Option<Span>, f: F)
+        where F: FnOnce(&mut Self)
+    {
+        let old_outer_impl_trait = self.outer_impl_trait;
```
Comment (Centril) [inline src/librustc_passes/ast_validation.rs]: use `mem::replace` here also?

**Comment by cramertj** [comment]
r=me with @centril's nits fixed.

File: src/librustc_passes/ast_validation.rs [hunk 6]
```diff
@@ -490,148 +595,10 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
...
-        f(self);
-        self.is_banned = old_is_banned;
-    }
-}
-
-impl<'a> Visitor<'a> for ImplTraitProjectionVisitor<'a> {
-    fn visit_ty(&mut self, t: &'a Ty) {
-        match t.node {
-            TyKind::ImplTrait(..) => {
-                if self.is_banned {
```
Comment (Zoxc) [inline src/librustc_passes/ast_validation.rs]: This PR will give errors about nested impl traits being banned in path parameters, unlike the old code, which doesn't visit inside impl Trait. So we'd get 3 errors then, one for nested impl Trait and one for each instance of impl Trait in a path parameter.

Comment (cramertj) [reply src/librustc_passes/ast_validation.rs]: SGTM.

**Pull request merged**